Backport of the following upstream commit...

commit 74e2bd1fa3ae9695af566ad5a7a288898787b909
Author: Wey-Yi Guy <wey-yi.w.guy@intel.com>
Date:   Wed Feb 3 09:28:55 2010 -0800

    mac80211: tear down all agg queues when restart/reconfig hw
    
    When there is a need to restart/reconfig hw, tear down all the
    aggregation queues and let the mac80211 and driver get in-sync to have
    the opportunity to re-establish the aggregation queues again.
    
    Need to wait until driver re-establish all the station information before tear
    down the aggregation queues, driver(at least iwlwifi driver) will reject the
    stop aggregation queue request if station is not ready. But also need to make
    sure the aggregation queues are tear down before waking up the queues, so
    mac80211 will not sending frames with aggregation bit set.
    
    Signed-off-by: Wey-Yi Guy <wey-yi.w.guy@intel.com>
    Signed-off-by: John W. Linville <linville@tuxdriver.com>

diff -up linux-2.6.33.noarch/net/mac80211/util.c.orig linux-2.6.33.noarch/net/mac80211/util.c
--- linux-2.6.33.noarch/net/mac80211/util.c.orig	2010-02-24 13:52:17.000000000 -0500
+++ linux-2.6.33.noarch/net/mac80211/util.c	2010-04-13 11:49:24.000000000 -0400
@@ -1145,6 +1145,14 @@ int ieee80211_reconfig(struct ieee80211_
 		}
 	}
 
+	rcu_read_lock();
+	if (hw->flags & IEEE80211_HW_AMPDU_AGGREGATION) {
+		list_for_each_entry_rcu(sta, &local->sta_list, list) {
+			ieee80211_sta_tear_down_BA_sessions(sta);
+		}
+	}
+	rcu_read_unlock();
+
 	/* add back keys */
 	list_for_each_entry(sdata, &local->interfaces, list)
 		if (netif_running(sdata->dev))
